sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -341,6 +341,16 @@ trait JoinSelectionHelper {
     )
   }
 
+  def getBroadcastNestedLoopJoinBuildSide(hint: JoinHint): Option[BuildSide] = {
+    if (hintToNotBroadcastAndReplicateLeft(hint)) {
+      Some(BuildRight)
+    } else if (hintToNotBroadcastAndReplicateRight(hint)) {
+      Some(BuildLeft)
+    } else {
+      None
+    }
+  }
```

Comment:
This logic is saying that if we hint to NOT broadcast left, we can broadcast right, and vice-versa. Shouldn't we check the hint on the other side as well? 
i.e.
```
if (hintToNotBroadcastAndReplicateLeft(hint) && !hintToNotBroadcastAndReplicateRight(hint)) {
  Some(BuildRight)
} else if (hintToNotBroadcastAndReplicateRight(hint) && !hintToNotBroadcastAndReplicateLeft(hint)) {
  Some(BuildLeft)
} else {
  None
}
```

Reply:
At the moment, the new hint can be only set on one side and never on both. BNLJ is considered as the default join strategy and having no broadcast and replicate hints on both sides would mean there is no matching join strategy to use. If we were to adapt the method above, we can't keep the existing default logic that picks the broadcast side based on size (that could cause a correctness problem). What about adding validation the new hint is set only on one side?

Reply:
`What about adding validation the new hint is set only on one side?`
Sounds good.

Reply:
Cool, I will work on it. Created SPARK-44148.